Part 2 of changes for SQL Compatible Null Handling

nishantmonu51 opened the pull request:
Part 2 of changes for SQL Compatible Null Handling - 
This PR contains changes to improve handling of null values in druid by treating nulls as missing values which don’t actually exist. This will make it more compatible to SQL standard and help with integration with other existing BI systems which support ODBC/JDBC. (Detailed description in proposal - https://github.com/druid-io/druid/issues/4349)

NOTE - NULL and empty strings are treated differently in entire druid query processing system.

## Nullability for Aggregators/Metrics

As per the current implementation, most aggregators are coded to replace null values with default e.g. sum treats them as zeroes, min treats them as positive infinity, etc.
To match the new semantics and make aggregators nullable, where if an aggregator encounters only the null values the result will be null value following changes are made -

* Added a NullableAggregator/NullableBufferAggregator which is be used as a decorator for existing aggregator implementations to make them nullable. This will return null values if all the values aggregated are null, If any of the aggregated value is non-null, it will have same behavior as the delegate aggregator. 
* If new null behavior is enabled, aggregator factories will decorate the aggregators with NullableAggregator/NullableBufferAggregator. 
* Added a new abstract class NullableAggregatorFactory which helps in reducing duplicate code and decide to switch between nullable and non-nullable aggregators. 
* Cardinality and HyperUnique  aggregators ignore null values and only count non-null values, If all the encountered values are null, the result will be 0. This is different from current behavior where null values are also counted.
* Count aggregator - since count aggregator is not associated with any column and just a count of the number of rows, it has the same behavior as before. 

## Math Expressions Null Handling
* For general calculations like sum, full expression is be considered as null if any of the components is null. 
* StringLiteral now supports null and and expressions containing non-quoted null is parsed as StringLiteral with null 
* Specifying a default value for null is supported by the use of NVL or IF clause to assign default values at query time.
* Comparisons between null and non-null values will follow three-valued logic and the result of a comparison involving NULL is not a boolean value – it is a null value.

## Filtering on Null values
* SelectorDimFilter currently specifies filtering on null values but the implementation assumes null and empty strings as equivalent. The implementation is changed to consider null and empty string differently. Generation of cache key for selectorDimFilter is also modified to support null.
* InFilter/ExpressionFilter/LikeFilter are also modified to not treat nulls as empty strings. 

## Changes to Druid build-in SQL layer
* NULL and empty strings are treated differently. 
* IS NULL and IS NOT NULL filters against null values instead of empty strings. 
* NVL, IFNULL and COALESCE work as per SQL standard. 
* count now skips null values and works as per SQL standard. 

## Backwards Compatibility 
* **_"druid.generic.useDefaultValueForNull"_** is added as a new config. The default value is set to true for backwards compatibility. 
*  code places in query/ingestion where we converted emptyToNull or nullToEmpty now use NullHandlingHelper which only does the conversion if backwards compatibility is enabled. 
* Storage layer changes are backwards compatible for GenericIndexed/String columns. 
* For Numeric columns new V2 serde implementations are introduced which also supports serde of nullRowsBitmaps. Previous implementations are kept unchanged to read any existing segments. 

## Testing 
* Most of the unit tests that tested nulls are modified to have two branches of expected results to test new/old behavior both. 
* For CI have added new runs to test matrix to test new behavior for nulls. 
* To run the tests with new behavior run **_"-Ddruid.generic.useDefaultValueForNull=false"_**

Pending work - 
https://github.com/apache/incubator-druid/labels/Area%20-%20Null%20Handling

nishantmonu51 commented:
@leventov : Created a new PR, please review.

leventov commented:
@nishantmonu51 as far as I can see you didn't address all comments that I left in the last batch in the previous PR, please answer to them there (if you are able to open that PR)

nishantmonu51 commented:
@leventov: last batch of comments was re lines more than 120 cols, add line breaks and reformatted code. Also replied to your comments on previous PR, If you have any more comments can you please add them here ?

jihoonson commented:
Since this PR is tagged as `Design Review`, I'm going to review this PR too.

jihoonson commented:
@nishantmonu51 is this the last PR for https://github.com/apache/incubator-druid/issues/4349? Would you let me know which are not done yet in https://github.com/apache/incubator-druid/pull/5278 and this PR?

jihoonson commented:
Some comments and questions for the PR description.

> Aggrgator/BufferAggregator - Added a new method boolean isNull() which returns false by default. aggregators that support nullability can choose to override this and return true if the aggregated result is null.

This should be moved to https://github.com/apache/incubator-druid/pull/5278.

> Changes to Druid build-in SQL layer
> * NULL and empty strings are treated differently.

I haven't checked through the whole patch, so am wondering this is true only for the SQL layer or the entire Druid query processing system.

jihoonson commented on extensions-core/histogram/src/main/java/io/druid/query/aggregation/histogram/ApproximateHistogramAggregator.java:
```diff
@@ -59,7 +60,9 @@ public ApproximateHistogramAggregator(
   public void aggregate()
   {
-    histogram.offer(selector.getFloat());
+    if (NullHandling.replaceWithDefault() || !selector.isNull()) {
```
What does `NullHandling.replaceWithDefault()` mean here? Does that mean nulls have replaced with some default value at ingestion time? If so, I guess `select.isNull()` always return `false` and thus this check can be simplified into `if (!selector.isNull())`.

nishantmonu51 replied:
this means whether nulls should be replaced with default value. 
In case of ExpressionColumnValueSelector isNull will compute the expression and then give the result, the check is there to not have any performance impact of calling isNull for default case.

jihoonson replied:
Got it. Thanks. Would you add a comment here?

nishantmonu51 replied:
sure.

jihoonson commented on extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/LoadingLookup.java:
```diff
@@ -62,15 +63,19 @@ public LoadingLookup(
+  public String apply(@Nullable final String key)
   {
-    if (key == null) {
+    String keyEquivalent = NullHandling.nullToEmptyIfNeeded(key);
```
I thought this kind of null value conversion is needed only when retrieving values from the storage layer (like in `ColumnValueSelector`), so that we expect `key` is already converted to some default value if needed. But, apparently it seems not. Would you explain why this is needed after retrieving values?

nishantmonu51 replied:
this is to preserve backwards compatibility, 
previously, nulls and empty strings were considered same. 
and if a looking has mapping "" -> "someval" then all nulls/empty strings will be resolved to "someval".

jihoonson replied:
Thanks.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/Expr.java:
```diff
@@ -365,15 +371,21 @@ public ExprEval eval(ObjectBinding bindings)
     if (leftVal.type() == ExprType.STRING && rightVal.type() == ExprType.STRING) {
       return evalString(leftVal.asString(), rightVal.asString());
     } else if (leftVal.type() == ExprType.LONG && rightVal.type() == ExprType.LONG) {
+      if (NullHandling.sqlCompatible() && (leftVal.isNumericNull() || rightVal.isNumericNull())) {
```
https://github.com/apache/incubator-druid/issues/5991

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -245,36 +252,63 @@ public final ExprType type()
-      return theLong == null ? 0L : theLong;
+      Number number = asNumber();
+      if (number == null) {
+        assert NullHandling.replaceWithDefault();
```
IIRC, I was actually doing that before, but changed this to assert based on discussions on previous PRs.

leventov replied:
Yes

jihoonson replied:
So, would you point me out where the discussion is? I guess it's in https://github.com/apache/incubator-druid/pull/5278?

nishantmonu51 replied:
unable to find the discussion among lots of review comments. 
I believe the intent here was to catch any coding errors/places where we may have missed handling nulls properly.
@leventov: can probably add more context here.

leventov replied:
https://github.com/apache/incubator-druid/pull/5278#discussion_r166700195 https://github.com/apache/incubator-druid/pull/5278#discussion_r167577131 https://github.com/apache/incubator-druid/pull/5278#discussion_r167579971 https://github.com/apache/incubator-druid/pull/5278#discussion_r167626205

nishantmonu51 commented:
Thanks @jihoonson for helping with the review and taking this PR forward.

leventov commented on processing/src/main/java/io/druid/query/aggregation/first/FloatFirstAggregatorFactory.java:
```diff
@@ -32,19 +32,21 @@
 import java.util.Objects;
 
-public class FloatFirstAggregatorFactory extends AggregatorFactory
+public class FloatFirstAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector>
```
This comment: https://github.com/apache/incubator-druid/pull/5452#discussion_r200325687 is not addressed.

nishantmonu51 replied:
not a blocker for this PR, would like to do it a follow up PR - https://github.com/apache/incubator-druid/issues/6039

leventov commented on processing/src/main/java/io/druid/query/expression/ExprUtils.java:
```diff
@@ -73,7 +74,14 @@ public static PeriodGranularity toPeriodGranularity(
       Chronology chronology = timeZone == null ? ISOChronology.getInstanceUTC() : ISOChronology.getInstance(timeZone);
       final Object value = originArg.eval(bindings).value();
-      origin = value != null ? new DateTime(value, chronology) : null;
+      if (value instanceof String && StringUtils.isEmpty((String) value)) {
```
This comment: https://github.com/apache/incubator-druid/pull/5452#discussion_r200339008 is not addressed

nishantmonu51 replied:
https://github.com/apache/incubator-druid/issues/6040
Will fix this line in this PR and prohibit the api in future PR.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -90,7 +95,7 @@ public SelectorDimFilter(
   public DimFilter optimize()
   {
-    return new InDimFilter(dimension, ImmutableList.of(value), extractionFn).optimize();
+    return new InDimFilter(dimension, Arrays.asList(value), extractionFn).optimize();
```
Collections.singletonList()

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/filter/ValueGetter.java:
```diff
@@ -27,5 +29,7 @@
   // converted to strings. We should also add functions
   // for these and modify ColumnComparisonFilter to handle
   // comparing Long and Float columns to eachother.
+  // Returns null when the underlying Long/Float value is null.
```
Not addressed: https://github.com/apache/incubator-druid/pull/5452#discussion_r200339349

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/groupby/orderby/DefaultLimitSpec.java:
```diff
@@ -269,12 +270,22 @@ public int compare(Row left, Row right)
   private Ordering<Row> metricOrdering(final String column, final Comparator comparator)
   {
-    return Ordering.from(Comparator.comparing((Row row) -> row.getRaw(column), Comparator.nullsLast(comparator)));
+    if (NullHandling.sqlCompatible()) {
```
Need a comment explaining this.

nishantmonu51 replied:
added comment.

leventov commented on processing/src/main/java/io/druid/query/groupby/orderby/DefaultLimitSpec.java:
```diff
@@ -269,12 +270,22 @@ public int compare(Row left, Row right)
-    return Ordering.from(Comparator.comparing((Row row) -> row.getDimension(dimension).isEmpty() ? null : row.getDimension(dimension).get(0), Comparator.nullsFirst(comparator)));
+    return Ordering.from(
+        Comparator.comparing(
+            (Row row) -> row.getDimension(dimension).isEmpty()
```
Don't call `row.getDimension(dimension)` twice. Please extract a small class, don't inline. It's unreadable.

nishantmonu51 replied:
extracted method to avoid calling getDimension twice.

leventov commented on processing/src/main/java/io/druid/query/topn/TopNMapFn.java:
```diff
@@ -56,14 +56,10 @@
-    final Double doubleValue = DimensionHandlerUtils.convertObjectToDouble(input);
-    return doubleValue == null ? DimensionHandlerUtils.ZERO_DOUBLE : doubleValue;
-  };
+  private static Function<Object, Object> FLOAT_TRANSFORMER = input -> DimensionHandlerUtils.convertObjectToFloat(input);
```
Could be a method reference?

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/topn/TopNMapFn.java:
```diff
@@ -56,14 +56,10 @@
-  };
+  private static Function<Object, Object> FLOAT_TRANSFORMER = input -> DimensionHandlerUtils.convertObjectToFloat(input);
+
+  private static Function<Object, Object> DOUBLE_TRANSFORMER = input -> DimensionHandlerUtils.convertObjectToDouble(
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/virtual/ExpressionColumnValueSelector.java:
```diff
@@ -82,6 +82,6 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
   public boolean isNull()
   {
-    return getObject().isNull();
+    return getObject().isNumericNull();
```
Why this?

nishantmonu51 replied:
It is possible for an expression to have a non-null String value but it can return null when parsed as a primitive long/float/double.
   ExprEval.isNumericNull checks whether the parsed primitive value is null or not.

nishantmonu51 replied:
added comment to code.

leventov commented on processing/src/main/java/io/druid/segment/virtual/ExpressionSelectors.java:
```diff
@@ -248,23 +248,32 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
         // Unknown ValueType. Try making an Object selector and see if that gives us anything useful.
-        supplier = supplierFromObjectSelector(columnSelectorFactory.makeColumnValueSelector(columnName));
+        supplier = supplierFromObjectSelector(columnSelectorFactory
+                .makeColumnValueSelector(columnName));
```
Not formatted properly

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/virtual/SingleStringInputCachingExpressionColumnValueSelector.java:
```diff
@@ -139,7 +139,7 @@ private ExprEval eval()
   public boolean isNull()
   {
-    return eval().isNull();
+    return eval().isNumericNull();
```
Why this?

nishantmonu51 replied:
added comment to code.

leventov commented on processing/src/main/java/io/druid/segment/virtual/SingleLongInputCachingExpressionColumnValueSelector.java:
```diff
@@ -150,6 +150,6 @@ private ExprEval eval(final long value)
   public boolean isNull()
   {
-    return getObject().isNull();
+    return getObject().isNumericNull();
```
Why this?

nishantmonu51 replied:
It is possible for an expression to have a non-null String value but it can return null when parsed as a primitive long/float/double.
ExprEval.isNumericNull checks whether the parsed primitive value is null or not.
Added comment to code.

leventov commented on server/src/main/java/io/druid/server/listener/announcer/ListeningAnnouncerConfig.java:
```diff
@@ -93,7 +93,7 @@ public String getAnnouncementPath(String listenerName)
   {
     return ZKPaths.makePath(
         getListenersPath(), Preconditions.checkNotNull(
```
Each argument should be on a separate line

nishantmonu51 replied:
done.

leventov commented on sql/src/main/java/io/druid/sql/calcite/planner/Calcites.java:
```diff
@@ -107,6 +108,7 @@ public static SchemaPlus createRootSchema(final Schema druidSchema, final Author
   public static String escapeStringLiteral(final String s)
   {
+    Preconditions.checkNotNull(s);
     if (s == null) {
```
This check immediately after Preconditions.checkNotNull?

nishantmonu51 replied:
removed the == null check, it was not needed.

leventov commented on sql/src/main/java/io/druid/sql/calcite/rel/DruidSemiJoin.java:
```diff
@@ -294,7 +295,11 @@ public RelOptCost computeSelfCost(final RelOptPlanner planner, final RelMetadata
               final Object value = row[i];
-              final String stringValue = value != null ? String.valueOf(value) : "";
+              if (value == null) {
+                // NULLS are not supposed to match NULLs in a join. So ignore them.
```
NULL*s*

nishantmonu51 replied:
done.

leventov commented on sql/src/main/java/io/druid/sql/calcite/planner/DruidRexExecutor.java:
```diff
@@ -84,7 +84,7 @@ public void reduce(
           literal = rexBuilder.makeLiteral(exprResult.asBoolean(), constExp.getType(), true);
         } else if (sqlTypeName == SqlTypeName.DATE) {
-          if (!constExp.getType().isNullable() && exprResult.isNull()) {
+          if (!constExp.getType().isNullable() && exprResult.isNumericNull()) {
```
Why this?

nishantmonu51 replied:
It is possible for an expression to have a non-null String value but it can return null when parsed as a primitive long/float/double.
ExprEval.isNumericNull checks whether the parsed primitive value is null or not.
Added comment to code.

leventov commented on sql/src/main/java/io/druid/sql/calcite/rel/DruidSemiJoin.java:
```diff
@@ -308,16 +313,18 @@ public RelOptCost computeSelfCost(final RelOptPlanner planner, final RelMetadata
-                        getCluster().getRexBuilder().makeLiteral(value)
-                    )
-                );
+                // NULLS are not supposed to match NULLs in a join. So ignore them.
```
NULL*s*

nishantmonu51 replied:
done.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -99,10 +98,10 @@ public Object value()
     return value;
   }
 
-  public boolean isNull()
```
this method is replaced with isNumericNull as that is what we need to check for Expressions. 
isNumericNull will check whether the primitive long/float/double equivalent is null or not. 
It is possible for an expression to have a non-null String value but it can return null when parsed as a primitive long/float/double.

nishantmonu51 commented:
@leventov @jihoonson : Thanks for the review comments, addressed them. 
Do you have any more comments ? 
Also, can you please add some info on how much of the review is pending and when can we expect to merge this one in.
Its blocking feature development/patches for Apache Hive Integration for handling nulls coming from druid properly.

jihoonson commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,114 @@
+ * Support Nullable Aggregations are encouraged to extend this class.
+ */
+@ExtensionPoint
+public abstract class NullableAggregatorFactory<T extends BaseNullableColumnValueSelector> extends AggregatorFactory
```
Currently only numeric aggregatorFactories extend this class. Do you have a plan for making other aggregatorFactories nullable?

jihoonson replied:
Thanks.

jihoonson commented on processing/src/main/java/io/druid/query/aggregation/post/DoubleGreatestPostAggregator.java:
```diff
@@ -81,13 +77,15 @@ public Comparator getComparator()
-          retVal = other;
+    Double retVal = NullHandling.replaceWithDefault() ? Double.NEGATIVE_INFINITY : null;
+    while (fieldsIter.hasNext()) {
+      Number nextVal = ((Number) fieldsIter.next().compute(values));
```
It looks that we need specialized `PostAggregator` for primitive types and a new method `isNull()` to avoid boxing/unboxing. Would you file an issue for this?

leventov replied:
Why? Does performance matter here?

jihoonson replied:
I think so. `compute()` method is called per row on brokers.

jihoonson commented on processing/src/main/java/io/druid/query/extraction/StrlenExtractionFn.java:
```diff
@@ -40,6 +41,9 @@ public static StrlenExtractionFn instance()
   public String apply(@Nullable String value)
   {
+    if (NullHandling.sqlCompatible() && value == null) {
+      return null;
```
Please annotate `@Nullable`.

nishantmonu51 replied:
done.

jihoonson commented on processing/src/main/java/io/druid/query/filter/InDimFilter.java:
```diff
@@ -121,21 +116,30 @@ public ExtractionFn getExtractionFn()
     byte[] extractionFnBytes = extractionFn == null ? new byte[0] : extractionFn.getCacheKey();
 
-    ByteBuffer filterCacheKey = ByteBuffer.allocate(3
+    ByteBuffer filterCacheKey = ByteBuffer.allocate(5
```
nit: can we use `CacheKeyBuilder` here?

nishantmonu51 replied:
done.

jihoonson commented on processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java:
```diff
@@ -77,10 +81,12 @@ public SelectorDimFilter(
     byte[] extractionFnBytes = extractionFn == null ? new byte[0] : extractionFn.getCacheKey();
 
-    return ByteBuffer.allocate(3 + dimensionBytes.length + valueBytes.length + extractionFnBytes.length)
+    return ByteBuffer.allocate(5 + dimensionBytes.length + valueBytes.length + extractionFnBytes.length)
```
nit: can we use `CacheKeyBuilder` here?

nishantmonu51 replied:
done.

jihoonson commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -1700,6 +1712,78 @@ public BufferComparator getBufferComparator()
     }
+
+    // This class is only used when SQL compatible null handling is enabled.
+    private class NullableRowBasedKeySerdeHelper implements RowBasedKeySerdeHelper
```
Please add a comment about the memory layout.

nishantmonu51 replied:
done.

jihoonson commented on processing/src/main/java/io/druid/segment/filter/LikeFilter.java:
```diff
@@ -90,7 +90,10 @@ public boolean supportsSelectivityEstimation(
     if (isSimpleEquals()) {
       // Verify that dimension equals prefix.
-      return ImmutableList.of(selector.getBitmapIndex(dimension, likeMatcher.getPrefix()));
+      return ImmutableList.of(selector.getBitmapIndex(
```
format

```java
      return ImmutableList.of(
          selector.getBitmapIndex(
              dimension,
              NullHandling.emptyToNullIfNeeded(likeMatcher.getPrefix())
          )
      );
```

nishantmonu51 replied:
done.

jihoonson commented on sql/src/main/java/io/druid/sql/calcite/expression/builtin/CeilOperatorConversion.java:
```diff
@@ -80,7 +80,7 @@ public DruidExpression toDruidExpression(
       return DruidExpression.fromFunctionCall(
           "timestamp_ceil",
-          ImmutableList.of(
+          Arrays.asList(
```
Why this change?

jihoonson replied:
Maybe `Stream.of()` is better.

leventov replied:
Changed because `ImmutableList` doesn't support nulls, I suppose. Anyway, a parallel PR #5980 handles this in the whole codebase.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java:
```diff
@@ -168,7 +168,7 @@ public static BaseFloatColumnValueSelector makeColumnValueSelectorWithFloatDefau
         {
           final ExprEval exprEval = baseSelector.getObject();
-          return exprEval.isNull() ? nullValue : (float) exprEval.asDouble();
+          return exprEval.isNumericNull() ? nullValue : (float) exprEval.asDouble();
```
can you elaborate on when it can be null, 
I think it will be ExprEval.of(null) instead of null itself.

jihoonson replied:
Ah, I just saw the return value of `BaseObjectColumnValueSelector.getObject()` is nullable. Looks that it's not in this case. Please add a comment about this.

nishantmonu51 replied:
Added comment.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/DoubleMaxAggregatorFactory.java:
```diff
@@ -51,25 +53,44 @@ public DoubleMaxAggregatorFactory(String name, String fieldName)
-    return new DoubleMaxAggregator(getDoubleColumnSelector(metricFactory, Double.NEGATIVE_INFINITY));
+    return getDoubleColumnSelector(
+        metricFactory,
+        Double.NEGATIVE_INFINITY
```
I believe that's because we use this as default when aggregating, No plan to change this, nullability is handled in NullableAggregatorFactory which encapsulates this.

jihoonson replied:
Ok, thanks.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregator.java:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  @Override
+  public boolean isNull()
```
yeah, one more reason to avoid enforcing calling isNull always was performance concern,
for some selectors it would lead to doing computations twice, 
e.g. https://github.com/apache/incubator-druid/issues/5169
We can always call isNull first and simplify the codepath once this is addressed.

jihoonson replied:
Thanks.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregator.java:
```diff
@@ -0,0 +1,107 @@
+  @Override
+  public boolean isNull()
+  {
+    return isNullResult || delegate.isNull();
```
there can be custom aggregators that result in null even when input is not null.

jihoonson replied:
Got it, thanks.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,114 @@
+   * @see AggregateCombiner
+   * @see io.druid.segment.IndexMerger
+   */
+  protected abstract AggregateCombiner makeAggregateCombiner2();
```
do you have any suggestions for a better name ? 
the original method is final in this class and adds nullability support for the combiner. subclasses will implement this method.

jihoonson replied:
It looks that this method is same with `makeAggregateCombiner()` before this PR. I think we can add a new method `makeNullableAggregateCombiner` and use it instead where `makeAggregateCombiner` is being used currently. For example, 

```java
  public AggregateCombiner makeNullableAggregateCombiner()
  {
    AggregateCombiner combiner = makeAggregateCombiner();
    return NullHandling.replaceWithDefault() ? combiner : new NullableAggregateCombiner(combiner);
  }
```

```java
...
  RowCombiningTimeAndDimsIterator(
      List<TransformableRowIterator> originalIterators,
      AggregatorFactory[] metricAggs,
      List<String> metricNames
  )
  {
...
    Arrays.setAll(combinedMetricSelectors, metricIndex -> metricAggs[metricIndex].makeNullableAggregateCombiner());
  }
```

nishantmonu51 replied:
thanks for the suggestion. done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableAggregatorFactory.java:
```diff
@@ -0,0 +1,114 @@
+   *
+   * @return the maximum number of bytes that an aggregator of this type will require for intermediate result storage.
+   */
+  protected abstract int getMaxIntermediateSize2();
```
any suggestions ?

jihoonson replied:
Similar to the above comment.

nishantmonu51 replied:
done.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,116 @@
+  @Override
+  public boolean isNull(ByteBuffer buf, int position)
+  {
+    return buf.get(position) == NullHandling.IS_NULL_BYTE || delegate.isNull(buf, position + Byte.BYTES);
```
this is to safeguard against any custom aggregators that result in null even when input is not null.

jihoonson replied:
Ok.

nishantmonu51 commented on extensions-core/lookups-cached-single/src/test/java/io/druid/server/lookup/LoadingLookupTest.java:
```diff
@@ -43,13 +44,20 @@
+    if (NullHandling.sqlCompatible()) {
+      // empty string should also have same behavior
+      Assert.assertEquals(null, loadingLookup.apply(""));
+    }
```
done.

nishantmonu51 commented on extensions-core/lookups-cached-single/src/test/java/io/druid/server/lookup/LoadingLookupTest.java:
```diff
@@ -43,13 +44,20 @@
+    if (NullHandling.sqlCompatible()) {
+      Assert.assertEquals(Collections.emptyList(), loadingLookup.unapply(null));
+    } else {
+      Assert.assertEquals(null, loadingLookup.unapply(null));
```
done.

nishantmonu51 commented on extensions-core/lookups-cached-single/src/test/java/io/druid/server/lookup/LoadingLookupTest.java:
```diff
@@ -68,7 +76,10 @@ public void testUnapplyAll() throws ExecutionException
-    Assert.assertEquals(ImmutableMap.of("value", Lists.newArrayList("key")), loadingLookup.unapplyAll(ImmutableSet.<String>of("value")));
+    Assert.assertEquals(
+        ImmutableMap.of("value", Lists.newArrayList("key")),
+        loadingLookup.unapplyAll(ImmutableSet.<String>of("value"))
```
removed.

nishantmonu51 commented on indexing-service/src/test/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTaskTest.java:
```diff
@@ -1466,7 +1505,7 @@ public void close()
   }
 
-  public long sumMetric(final Task task, final DimFilter filter, final String metric)
+  public Long sumMetric(final Task task, final DimFilter filter, final String metric)
```
added.

nishantmonu51 commented on indexing-service/src/test/java/io/druid/indexing/common/task/RealtimeIndexTaskTest.java:
```diff
@@ -1081,7 +1088,7 @@ public void close()
   }
 
-  public long sumMetric(final Task task, final DimFilter filter, final String metric)
+  public Long sumMetric(final Task task, final DimFilter filter, final String metric)
```
done.

nishantmonu51 commented:
> Aggrgator/BufferAggregator - Added a new method boolean isNull() which returns false by default. aggregators that support nullability can choose to override this and return true if the aggregated result is null.
This should be moved to #5278.

Removed it from this PR. 

> Changes to Druid build-in SQL layer
NULL and empty strings are treated differently.
I haven't checked through the whole patch, so am wondering this is true only for the SQL layer or the entire Druid query processing system.

it applies to whole query processing. updated PR description to add a note.

nishantmonu51 commented:
@jihoonson : For the pending work, I have labelled the issues with Area : Null Handling 
https://github.com/apache/incubator-druid/labels/Area%20-%20Null%20Handling

jihoonson commented:
Thanks @nishantmonu51. Would you please check this comment: https://github.com/apache/incubator-druid/pull/5958#discussion_r204937099?

Also, I'm not sure what have done in https://github.com/apache/incubator-druid/pull/5278 + this PR and not. I think at least `Comparisons between null and non-null values will follow three-valued logic and the result of a comparison involving NULL is not a boolean value – it is a null value.` is not done yet? Would you please clarify what are remaining in https://github.com/apache/incubator-druid/issues/4349?

nishantmonu51 commented:
> Comparisons between null and non-null values will follow three-valued logic and the result of a comparison involving NULL is not a boolean value – it is a null value.
this actually works now for Expressions, See BinaryEvalOpExprBase.eval, if any of the values being compared is null it returns null. 
Do you see any places where this is might not work ?

nishantmonu51 commented:
@jihoonson In terms of functionality, After this PR, all things mentioned in the proposal except following should be working - 
* last of @gianm comments (https://github.com/apache/incubator-druid/issues/6053). 
* Ability to specify default values at ingestion 
>  8. Users will be able to specify default values to replace NULL values at ingestion time.

* Handling of follow up comments perf related 
* Updating/Adding docs for the new behavior

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/NullableBufferAggregator.java:
```diff
@@ -0,0 +1,116 @@
+ * aggregator. Note that the delegate aggregator is not required to perform check for
+ * {@link BaseNullableColumnValueSelector#isNull()} on the selector as only non-null values will be passed to the
+ * delegate aggregator. This class is only used when SQL compatible null handling is enabled.
+ */
```
done.

nishantmonu51 commented:
@jihoonson any more comments here ?

jihoonson commented:
@nishantmonu51 awesome! Thanks for clarifying. Everything looks good except the travis failure. Would you check it please?

asdf2014 commented on extensions-core/lookups-cached-single/src/test/java/io/druid/server/lookup/LoadingLookupTest.java:
```diff
@@ -67,7 +82,14 @@ public void testUnapplyAll() throws ExecutionException
             .andReturn(Collections.singletonList("key"))
             .once();
     EasyMock.replay(reverseLookupCache);
+<<<<<<< HEAD
```
Please resolve the conflict..

nishantmonu51 replied:
done.

nishantmonu51 commented:
@jihoonson: travis is passing now.

nishantmonu51 commented:
@leventov @jihoonson @b-slim Thanks for the review. 
This PR is ready for merge and has got required +1s. 
can one of you please merge this ?
